vcd/resource_vcd_vapp_vm.go

```diff
@@ -819,52 +836,64 @@ func resourceVcdVAppVmUpdate(d *schema.ResourceData, meta interface{}) error {
 	vcdClient.lockParentVapp(d)
 	defer vcdClient.unLockParentVapp(d)
 
-	return resourceVcdVAppVmUpdateExecute(d, meta)
-}
-
-func resourceVcdVAppVmUpdateExecute(d *schema.ResourceData, meta interface{}) error {
-	log.Printf("[DEBUG] [VM update] started without lock")
-
 	// Exit early only if "network_dhcp_wait_seconds" is changed because this field only supports
 	// update so that its value can be written into statefile and be accessible in read function
 	if onlyHasChange("network_dhcp_wait_seconds", vappVmSchema, d) {
 		log.Printf("[DEBUG] [VM update] exiting early because only 'network_dhcp_wait_seconds' has change")
 		return resourceVcdVAppVmRead(d, meta)
 	}
 
-	vcdClient := meta.(*VCDClient)
-
-	org, vdc, err := vcdClient.GetOrgAndVdcFromResource(d)
+	err := resourceVmHotUpdate(d, meta)
 	if err != nil {
-		return fmt.Errorf(errorRetrievingOrgAndVdc, err)
+		return err
 	}
 
-	vapp, err := vdc.GetVAppByName(d.Get("vapp_name").(string), false)
+	return resourceVcdVAppVmUpdateExecute(d, meta, "update")
+}
 
+func resourceVmHotUpdate(d *schema.ResourceData, meta interface{}) error {
+	vcdClient, _, vdc, vapp, _, vm, err := getVmFromResource(d, meta)
 	if err != nil {
-		return fmt.Errorf("error finding vApp: %s", err)
+		return err
 	}
-
-	identifier := d.Id()
-	if identifier == "" {
-		identifier = d.Get("name").(string)
+	if d.Get("memory_hot_add_enabled").(bool) && d.HasChange("memory") {
+		err = changeMemorySize(d, vm)
+		if err != nil {
+			return err
+		}
 	}
-	if identifier == "" {
-		return fmt.Errorf("[VM update] neither name or ID was set")
+	if d.Get("cpu_hot_add_enabled").(bool) && d.HasChange("cpus") {
+		err = changeCpuCount(d, vm)
+		if err != nil {
+			return err
+		}
 	}
 
-	vm, err := vapp.GetVMByNameOrId(identifier, false)
+	if d.HasChange("network") && !isNetworkRemoved(d) {
```

Comment:
Please add a comment above on why we have special logic in case network was not removed.

Reply:
added